Add run network

icecrime opened the pull request:
Since we added support for top-level networks, it feels like `--network` would be a better name for the corresponding `docker run` flag.

This PR adds the `--network` flag and hides the `--net` one, without removing it.

Cc @dnephin @mavenugo.

Commit:
Organize run flags into categories

Flatten the list of `docker run` flags and group them loosely by
category (general purpose, security, networking, ...).

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

Commit:
Rename `--net` to `--network`

Add a `--network` flag which replaces `--net` without deprecating it
yet. The `--net` flag remains hidden and supported.

Add a `--network-alias` flag which replaces `--net-alias` without deprecating
it yet. The `--net-alias` flag remains hidden and supported.

Signed-off-by: Arnaud Porterie (icecrime) <arnaud.porterie@docker.com>

dnephin commented:
LGTM

thaJeztah commented:
this needs a deprecation announcement in https://github.com/docker/docker/blob/master/docs/deprecated.md

icecrime commented:
@thajeztah Do you think we should deprecate? I'm tempted not to given the popularity of `--net` and the fact that it existed for years.

Keeping it hidden and changing the docs just means that more people will start moving to `--network` over time.

WDYT?

thaJeztah commented:
@icecrime ah, alright, I interpreted it as "deprecating" (which, in a way it is). That's a first, perhaps we should document it there, but mention that both will stay supported (just that `--network` is preferred).

alright, LGTM we can discuss those changes later if needed

cpuguy83 commented:
There is also `--net-alias`

FrenchBen commented:
👍  to @cpuguy83's comment - If we're moving to a versbose `--network` it would be nice to change `--net-alias` to `--network-alias`

crosbymichael commented:
@icecrime can you rebase?

ibuildthecloud commented:
:-1: More to type and a pedantic change.  You'll never confused `net` for not meaning `network`.  If you were to ever actually remove `--net` I'll probably throw a tantrum so why keep a hidden setting forever. (Assuming you don't want me to throw a tantrum.)

icecrime commented:
Updated, rebased, added `--network-alias`.

@ibuildthecloud I like typing :innocent:

My preference goes to `--network` because it maps to `docker network` top-level command. The `--net` flag was introduced back when this corresponded to a network mode (such as `none`, or `host`). I think it's better explained to map with the corresponding top-level network object now that it's a thing.

I'll leave it to maintainers and @mavenugo.

FrenchBen commented on docs/reference/commandline/create.md:
```diff
@@ -62,13 +62,13 @@ Creates a new container.
       --memory-swappiness=""        Tune a container's memory swappiness behavior. Accepts an integer between 0 and 100.
       --name=""                     Assign a name to the container
-      --net="bridge"                Connect a container to a network
+      --net-alias=[]                Add network-scoped alias for the container
```
`--network-alias` ?

FrenchBen commented on docs/reference/commandline/run.md:
```diff
@@ -72,13 +72,13 @@ parent = "smn_cli"
       --memory-swappiness=""        Tune a container's memory swappiness behavior. Accepts an integer between 0 and 100.
       --name=""                     Assign a name to the container
-      --net="bridge"                Connect a container to a network
+      --net-alias=[]                Add network-scoped alias for the container
```
`--network-alias` ?

FrenchBen commented on docs/reference/run.md:
```diff
@@ -283,22 +283,22 @@ of the containers.
+                       'container:<name|id>': reuse another container's network stack
+                       'host': use the Docker host network stack
+                       '<network-name>|<network-id>': connect to a user-defined network
+    --net-alias=[]     : Add network-scoped alias for the container
```
`--network-alias` ?

jfrazelle commented:
sgtm, but I wouldn't depreciate `--net` I like just hiding it, mostly because it's sooooo prevalent

jfrazelle commented:
also I prefer `--network`

thaJeztah commented:
needs a rebase now

tiborvass commented:
@icecrime feel free to rebase :)

icecrime commented:
Rebased

cpuguy83 commented:
LGTM if janky is happy.

thaJeztah commented on runconfig/opts/parse.go:
```diff
@@ -446,99 +461,106 @@ func Parse(flags *pflag.FlagSet, copts *ContainerOptions) (*container.Config, *c
+		config.StopSignal = copts.flStopSignal
+	}
+
+	if flags.Changed("net") && flags.Changed("network") {
```
I was looking at the pflags package to see if it has an option to actually specify an *alias* for flags, and saw this; https://github.com/spf13/pflag#mutating-or-normalizing-flag-names. Would that be a cleaner solution?

thaJeztah replied:
Actually, i think I like the current approach more, ignore

thaJeztah commented on docs/reference/commandline/run.md:
```diff
@@ -78,13 +78,13 @@ Options:
       --name string                 Assign a name to the container
-      --network string              Connect a container to a network (default "default")
+      --network-alias=[]            Add network-scoped alias for the container
+      --network="bridge"            Connect a container to a network
```
Looks like this was reverted to the old output here

thaJeztah commented on docs/reference/commandline/create.md:
```diff
@@ -72,7 +72,7 @@ Options:
       --memory-swappiness int       Tune container memory swappiness (0 to 100) (default -1)
       --name string                 Assign a name to the container
-      --net string                  Connect a container to a network (default "default")
+      --network string              Connect a container to a network (default "default")
```
Looks like you forgot to update `--net-alias` -> `--network-alias` here

thaJeztah commented on runconfig/opts/parse.go:
```diff
@@ -446,99 +461,106 @@ func Parse(flags *pflag.FlagSet, copts *ContainerOptions) (*container.Config, *c
+		return nil, nil, nil, fmt.Errorf("--net and --network cannot be both set")
+	}
+	if flags.Changed("net-alias") && flags.Changed("network-alias") {
+		return nil, nil, nil, fmt.Errorf("--net-alias and --network-alias cannot be both set")
```
Not sure this is correct; I think this should be valid;

    docker run --net-alias=foo --network-alias=bar --name=baz foobar

thaJeztah replied:
Or would that require https://github.com/docker/docker/pull/23324#discussion_r70023639 to be used after all?

vdemeester replied:
I do agree with @thaJeztah :angel:

thaJeztah commented:
@icecrime couple of nits/questions

also, this will need updates to the bash and zsh (and fish) completion scripts; perhaps @albers and @sdurrheimer can take care of those if you want

vdemeester commented on runconfig/opts/parse.go:
```diff
@@ -446,99 +461,106 @@ func Parse(flags *pflag.FlagSet, copts *ContainerOptions) (*container.Config, *c
+	}
+
+	if flags.Changed("net") && flags.Changed("network") {
+		return nil, nil, nil, fmt.Errorf("--net and --network cannot be both set")
```
I'm not sure why this is needed. If `docker run --net mynetwork1 --net mynetwork2 …` is possible (only taking into account the latest one), then `docker run --net mynetwork1 --network mynetwork2 …` could/should be possible too :angel:.

thaJeztah replied:
I think it won't work with the current implementation because two flags (net/network) write to the same variable, so the outcome is "undefined".

vdemeester replied:
@thaJeztah not really, it's just the same as of now, the latest one wins (I tried it out :angel: ).

vdemeester commented on docs/reference/commandline/run.md:
```diff
@@ -78,13 +78,13 @@ Options:
       --name string                 Assign a name to the container
-      --net string                  Connect a container to a network (default "default")
+      --network-alias=[]            Add network-scoped alias for the container
+      --network="bridge"            Connect a container to a network
```
I'm not sure how to handle that, but as we do not deprecate `--net`, both flags should appear here (as this is, at some *really small bits*, the same output as `docker run --help`)

vdemeester replied:
Ah, didin't see `flags.MarkHidden("net")`, nevermind my comment :sweat_smile:

albers commented:
I'll update bash completion, thanks for the ping @thaJeztah

icecrime commented:
Rebased again (:cry:), and fixed the comments.

thaJeztah commented:
LGTM, thanks!

ping @cpuguy83 PTAL

vdemeester commented:
I think we can move this in `docs-review` then :angel: 
docs LGTM 🐱

thaJeztah commented:
docs LGTM as well

vdemeester commented:
@icecrime might be related ?

```
19:03:41 ----------------------------------------------------------------------
19:03:41 FAIL: docker_cli_network_unix_test.go:1664: DockerNetworkSuite.TestDockerNetworkFlagAlias
19:03:41 
19:03:41 docker_cli_network_unix_test.go:1670:
19:03:41     c.Assert(status, checker.Not(checker.Equals), 0, check.Commentf("unexpected status code %d (%s)", status, output))
19:03:41 ... obtained int = 0
19:03:41 ... expected int = 0
19:03:41 ... unexpected status code 0 ()
19:03:41 
19:04:20 
19:04:20 ----------------------------------------------------------------------
```

vdemeester commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -1660,3 +1660,17 @@ func (s *DockerDaemonSuite) TestDaemonRestartRestoreBridgeNetwork(t *check.C) {
+
+	output, status, _ = dockerCmdWithError("run", "--rm", "--net=user", "--network=user", "busybox", "true")
+	c.Assert(status, checker.Not(checker.Equals), 0, check.Commentf("unexpected status code %d (%s)", status, output))
+	c.Assert(status, checker.Equals, 0, check.Commentf("unexpected status code %d (%s)", status, output))
```
@icecrime I think this one shouldn't be there :sweat_smile:

vdemeester commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -1660,3 +1660,17 @@ func (s *DockerDaemonSuite) TestDaemonRestartRestoreBridgeNetwork(t *check.C) {
+
+	output, status, _ = dockerCmdWithError("run", "--rm", "--net-alias=foo", "--network-alias=bar", "busybox", "true")
+	c.Assert(status, checker.Not(checker.Equals), 0, check.Commentf("unexpected status code %d (%s)", status, output))
+	c.Assert(status, checker.Equals, 0, check.Commentf("unexpected status code %d (%s)", status, output))
```
Same here 👼

icecrime commented:
@vdemeester Oh my... fixed :innocent: 

Also rebased to fix documentation CI.

thaJeztah commented:
LOL, looks like there's issues with source forge now. I'll try to find alternative URLs, because sourceforge is evil anyway;

```
18:43:27 ERROR: Failed to crawl "http://lxc.sourceforge.net/index.php/about/kernel-namespaces/"  Get http://lxc.sourceforge.net/index.php/about/kernel-namespaces/: read tcp 172.17.0.2:43572->216.34.181.96:80: read: connection reset by peer
18:43:27 ERROR: (888) 1 links to (http://lxc.sourceforge.net/index.php/about/kernel-namespaces/)
18:43:27 			 link http://lxc.sourceforge.net/index.php/about/kernel-namespaces/ on page engine/security/security.md
18:43:53 ERROR: Failed to crawl "http://aufs.sourceforge.net/aufs3/man.html"  Get http://aufs.sourceforge.net/aufs3/man.html: read tcp 172.17.0.2:43615->216.34.181.96:80: read: connection reset by peer
18:43:53 ERROR: (888) 1 links to (http://aufs.sourceforge.net/aufs3/man.html)
18:43:53 			 link http://aufs.sourceforge.net/aufs3/man.html on page engine/reference/builder.md
```

thaJeztah commented:
re-LGTM (if CI is green)

vieux commented:
LGTM (if CI is green)

icecrime commented:
Unrelated failure in experimental (cc @thaJeztah)

```
13:36:18 FAIL: docker_cli_daemon_test.go:2095: DockerDaemonSuite.TestDaemonRestartWithKilledRunningContainer
13:36:18 
13:36:18 [d52686177] waiting for daemon to start
13:36:18 [d52686177] daemon started
13:36:18 [d52686177] waiting for daemon to start
13:36:18 [d52686177] daemon started
13:36:18 docker_cli_daemon_test.go:2144:
13:36:18     t.Fatalf("Expected exit code '%s' got '%s' for container '%s'\n", "143", out, cid)
13:36:18 ... Error: Expected exit code '143' got '0' for container '0d6788ddd70d2635e91e0c968adef1102f837f8989fb162ffa6f442d0565f615'
13:36:18 
13:36:18 
13:36:18 [d52686177] exiting daemon
```

thaJeztah commented:
@icecrime have an issue for that; https://github.com/docker/docker/issues/24203

tiborvass commented:
LGTM

wrp commented:
"You'll never confused net for not meaning network."  Completely untrue.  Here it is 7 years later, and the fact that --net is hidden in the docs caused me to spend (read "waste") several 10s of minutes tracking down whether there is if fact some subtle difference between --net and --network.  Either "unhide" --net and document it, or fully deprecate it.  This is not something that will "never cause confusion".  Rather, this is a land mine of confusion.  Deprecate it or document it.  Choose.
